pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java

```diff
@@ -0,0 +1,170 @@
+        throw new IllegalStateException(String.format("Partition id mismatched for %s, expected partition id: %d",
+            segmentName, partitionId));
+      }
+      partitionId = segmentPartitionId;
+      maxCreationTimeOfMergingSegments = Math.max(maxCreationTimeOfMergingSegments,
+          segmentMetadata.getIndexCreationTime());
+
+      String crcFromDeepStorageSegment = segmentMetadata.getCrc();
+      if (!originalSegmentCrcFromTaskGenerator.get(i).equals(crcFromDeepStorageSegment)) {
+        String message = String.format("Crc mismatched between ZK and deepstore copy of segment: %s. Expected crc "
+                + "from ZK: %s, crc from deepstore: %s", segmentName, originalSegmentCrcFromTaskGenerator.get(i),
+            crcFromDeepStorageSegment);
+        LOGGER.error(message);
+        throw new IllegalStateException(message);
+      }
+      RoaringBitmap validDocIds = MinionTaskUtils.getValidDocIdFromServerMatchingCrc(tableNameWithType, segmentName,
+              ValidDocIdsType.SNAPSHOT.name(), MINION_CONTEXT, crcFromDeepStorageSegment);
+      if (validDocIds == null) {
+        // no valid crc match found or no validDocIds obtained from all servers
+        // error out the task instead of silently failing so that we can track it via task-error metrics
+        String message = String.format("No validDocIds found from all servers. They either failed to download "
+                + "or did not match crc from segment copy obtained from deepstore / servers. " + "Expected crc: %s",
+            "");
+        LOGGER.error(message);
+        throw new IllegalStateException(message);
+      }
+
+      recordReaders.add(new CompactedPinotSegmentRecordReader(segmentDir, validDocIds));
+    }
+
+    segmentProcessorConfigBuilder.setSegmentNameGenerator(
+        new UploadedRealtimeSegmentNameGenerator(TableNameBuilder.extractRawTableName(tableNameWithType), partitionId,
+            System.currentTimeMillis(), MinionConstants.UpsertCompactMergeTask.MERGED_SEGMENT_NAME_PREFIX, null));
+    if (maxCreationTimeOfMergingSegments != 0) {
+      segmentProcessorConfigBuilder.setCustomCreationTime(maxCreationTimeOfMergingSegments);
+    }
+    SegmentProcessorConfig segmentProcessorConfig = segmentProcessorConfigBuilder.build();
+    List<File> outputSegmentDirs;
+    try {
+      _eventObserver.notifyProgress(_pinotTaskConfig, "Generating segments");
+      outputSegmentDirs = new SegmentProcessorFramework(recordReaders, segmentProcessorConfig, workingDir).process();
+    } finally {
+      for (RecordReader recordReader : recordReaders) {
+        recordReader.close();
+      }
+    }
+
+    long endMillis = System.currentTimeMillis();
+    LOGGER.info("Finished task: {} with configs: {}. Total time: {}ms", taskType, configs, (endMillis - startMillis));
+
+    List<SegmentConversionResult> results = new ArrayList<>();
+    for (File outputSegmentDir : outputSegmentDirs) {
```

Comment:
just to be sure, could one task generate multiple new segments or just one? If multiple segments could be created, then how to handle cases when some segments failed to get uploaded.

Reply:
Current implementation ensures generating one segment. But if you think about it even if we generate multiple segments, we don't need to think about rollback much. It's just that whatever segments got uploaded, validDocIds will be uploaded for those and the corresponding segments would be discarded as their docs will be invalid (assuming takeSnapshot has run once).

Reply:
> For M:N merging, I think the computation of alreadyMergedSegments might need some extensions, as you may want to retry some of the input segments, instead of skipping them if they are in some segment's custom map.

Yeah or we ensure always selecting complete 'x' segments while creating N segments such that sum over x = M. For the sake of simplicity, we are resolving this in task-generation itself currently ensuring to generate multiple X:1 subtasks instead.